runner/internal/shim/docker.go

```diff
@@ -66,21 +64,33 @@ func NewDockerRunner(dockerParams DockerParameters) (*DockerRunner, error) {
 	runner := &DockerRunner{
 		client:       client,
 		dockerParams: dockerParams,
-		state:        Pending,
+		tasks:        NewTaskStorage(),
 	}
 	return runner, nil
 }
 
 func (d *DockerRunner) Run(ctx context.Context, cfg TaskConfig) error {
+	task := NewTask(cfg)
+
+	// For legacy API compatibility, since LegacyTaskID is the same for all tasks
+	if task.ID == LegacyTaskID {
+		d.tasks.Delete(task.ID)
+	}
+
+	if ok := d.tasks.Add(task); !ok {
+		return tracerr.Errorf("task %s is already submitted", task.ID)
+	}
+
+	defer func() { _ = d.tasks.Update(task) }()
```

Comment:
Would it make sense to check for `false` here and log a message in case of errors?